{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import seaborn as sns\n",
    "import scipy \n",
    "from functools import reduce\n",
    "import sys\n",
    "\n",
    "\n",
    "file_dirname = os.path.dirname(os.path.abspath('__file__'))\n",
    "sys.path.insert(1,os.path.join(file_dirname,'..','statistics'))\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "from module_analysis_transposon_sites import (frequency_transposons,reads_per_transposon, transposon_density,\n",
    "                                            median_feature_essentials,median_feature_nonessentials,filter_low_and_biased_reads_genes,\n",
    "                                            reads_distribution_per_gene)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "#%% Import of dataframes output from the SATAY pipeline\n",
    "\n",
    "\n",
    "names_libraries={'wt':'data_wt_merged.xlsx','dbem1dbem3dnrp1':'data-enzo-dbem13dnrp1.xlsx'}\n",
    "data_library=[]\n",
    "for i in names_libraries.keys():\n",
    " data_library.append(pd.read_excel(file_dirname+'/../../datasets/'+names_libraries[i],index_col='Unnamed: 0',engine='openpyxl'))\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "source": [
    "#%% Removing the ADE2 and URA3 gene reads and insertions from the population\n",
    "\n",
    "for i in np.arange(0,len(data_library)):\n",
    "    data_library[i]=data_library[i][data_library[i].Standard_name != 'ADE2']\n",
    "    data_library[i]=data_library[i][data_library[i].Standard_name != 'URA3']\n",
    "\n",
    "data_library_pd=pd.concat(data_library,keys=names_libraries.keys(),sort=True)\n",
    "data_library_pd.fillna(0,inplace=True)"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "\n",
    "measures_satay=defaultdict(dict)\n",
    "\n",
    "for keys in names_libraries.keys():\n",
    "    measures_satay[keys][\"bp_per_tr\"]=frequency_transposons(data_library_pd.loc[keys])\n",
    "    measures_satay[keys][\"reads_per_tr\"]=reads_per_transposon(data_library_pd.loc[keys])\n",
    "    measures_satay[keys][\"tr_pr_100bp\"]=100*transposon_density(data_library_pd.loc[keys])\n",
    "    measures_satay[keys][\"tr_essentials_median\"]=median_feature_essentials(data_library_pd.loc[keys],feature=\"Ninsertions\")\n",
    "    measures_satay[keys][\"tr_nonessentials_median\"]=median_feature_nonessentials(data_library_pd.loc[keys],feature=\"Ninsertions\")\n",
    "    measures_satay[keys][\"reads_per_tr_essentials\"]=median_feature_essentials(data_library_pd.loc[keys],feature=\"Nreadsperinsrt\")\n",
    "    measures_satay[keys][\"reads_per_tr_nonessentials\"]=median_feature_nonessentials(data_library_pd.loc[keys],feature=\"Nreadsperinsrt\")\n",
    "\n",
    "measures_satay_pd=pd.DataFrame(measures_satay)\n",
    "measures_satay_pd"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "There are, in average 4.1387178564499365 transposons per each 100 basepairs\n",
      "There are, in average 3.524039374568914 transposons per each 100 basepairs\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "                                   wt  dbem1dbem3dnrp1\n",
       "bp_per_tr                   24.162072        28.376527\n",
       "reads_per_tr                75.231884         8.096154\n",
       "tr_pr_100bp                  4.138718         3.524039\n",
       "tr_essentials_median        18.000000        12.000000\n",
       "tr_nonessentials_median     24.000000        21.000000\n",
       "reads_per_tr_essentials     41.709330         4.860390\n",
       "reads_per_tr_nonessentials  79.500000         8.490196"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt</th>\n",
       "      <th>dbem1dbem3dnrp1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bp_per_tr</th>\n",
       "      <td>24.162072</td>\n",
       "      <td>28.376527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads_per_tr</th>\n",
       "      <td>75.231884</td>\n",
       "      <td>8.096154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tr_pr_100bp</th>\n",
       "      <td>4.138718</td>\n",
       "      <td>3.524039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tr_essentials_median</th>\n",
       "      <td>18.000000</td>\n",
       "      <td>12.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tr_nonessentials_median</th>\n",
       "      <td>24.000000</td>\n",
       "      <td>21.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads_per_tr_essentials</th>\n",
       "      <td>41.709330</td>\n",
       "      <td>4.860390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reads_per_tr_nonessentials</th>\n",
       "      <td>79.500000</td>\n",
       "      <td>8.490196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "metadata": {},
     "execution_count": 6
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "data_wt=data_library_pd.loc['wt'].copy()\n",
    "data_mutant=data_library_pd.loc['dbem1dbem3dnrp1'].copy()\n",
    "\n",
    "data_nrp1=data_mutant\n",
    "####### Transposon density vs genes ####################\n",
    "\n",
    "data_nrp1['tr-density']=data_nrp1['Ninsertions']/data_nrp1['Nbasepairs']\n",
    "data_wt['tr-density']=data_wt['Ninsertions']/data_wt['Nbasepairs']\n",
    "data_nrp1['reads-density']=data_nrp1['Nreads']/data_nrp1['Nbasepairs']\n",
    "data_wt['reads-density']=data_wt['Nreads']/data_wt['Nbasepairs']\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    "### Filtering data to compute median_feature_nonessentials\n",
    "\n",
    "data_filter_wt=filter_low_and_biased_reads_genes(data_wt)\n",
    "data_filter_dnrp1=filter_low_and_biased_reads_genes(data_nrp1)\n",
    "\n",
    "data_filter_dnrp1.fillna(0,inplace=True)\n",
    "data_filter_wt.fillna(0,inplace=True)\n",
    "\n",
    "data_filter_wt.index=data_wt['Standard_name']\n",
    "data_filter_dnrp1.index=data_nrp1[\"Standard_name\"]"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "#%% Plot transposon density (fig 1B Benoit) highlighting the centromere position\n",
    "\n",
    "#data_wt.index=data_wt['Standard_name']\n",
    "\n",
    "fig = plt.figure(figsize=(10,5))\n",
    "ax = fig.add_subplot(111)\n",
    "ax.plot(data_wt['tr-density'],'.',alpha=0.5,color='b')\n",
    "ax.set_ylabel('transposon density: tn/bp')\n",
    "ax.set_xlabel('genes')\n",
    "\n",
    "## annotated centromeres\n",
    "for i in  data_wt.index:\n",
    "\n",
    "    if type(data_wt.loc[i,'Feature_type'])!=str:\n",
    "        \n",
    "        if data_wt.loc[i,'Feature_type'].any() =='Centromere': \n",
    "            \n",
    "            ax.vlines(x=i,ymin=0,ymax=0.8,linestyles='--',alpha=0.3)\n",
    "            ax.text(x=i,y=0.6,s='centromere',rotation=90,fontsize=8)\n",
    "    else:\n",
    "\n",
    "         if data_wt.loc[i,'Feature_type'] =='Centromere': \n",
    "            \n",
    "            ax.vlines(x=i,ymin=0,ymax=0.8,linestyles='--',alpha=0.3)\n",
    "            ax.text(x=i,y=0.6,s='centromere',rotation=90,fontsize=8)\n",
    "    "
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "#%% Transposon density for comparing two libraries\n",
    "fig=plt.figure(figsize=(10,9))\n",
    "grid = plt.GridSpec(2, 1, wspace=0.0, hspace=0.0)\n",
    "ax = plt.subplot(grid[0,0])\n",
    "ax2 = plt.subplot(grid[1,0]) \n",
    "\n",
    "ax.set_xticks([])\n",
    "# for minor ticks\n",
    "ax.set_xticks([], minor=True)\n",
    "\n",
    "ax.plot(data_wt['tr-density'],\".\",alpha=0.5,color='b')\n",
    "ax.set_ylabel('transposon density: tn/bp')\n",
    "\n",
    "## annotated centromeres\n",
    "for i in data_wt.index:\n",
    "    \n",
    "    if type(data_wt.loc[i,'Feature_type'])!=str:\n",
    "        \n",
    "        if data_wt.loc[i,'Feature_type'].any() =='Centromere': \n",
    "            \n",
    "            ax.vlines(x=i,ymin=0,ymax=0.8,linestyles='--',alpha=0.3)\n",
    "            ax.text(x=i,y=0.6,s='centromere',rotation=90,fontsize=8)\n",
    "    else:\n",
    "\n",
    "         if data_wt.loc[i,'Feature_type'] =='Centromere': \n",
    "            \n",
    "            ax.vlines(x=i,ymin=0,ymax=0.8,linestyles='--',alpha=0.3)\n",
    "            ax.text(x=i,y=0.6,s='centromere',rotation=90,fontsize=8)\n",
    "\n",
    "ax2.plot(data_nrp1['tr-density'],\".\",alpha=0.5,color='orange')\n",
    "ax2.set_ylabel('transposon density: tn/bp mutant ')\n",
    "ax2.set_xlabel('genes')\n",
    "\n",
    "ax2.set_xticks([])\n",
    "# for minor ticks\n",
    "ax2.set_xticks([], minor=True)\n",
    "## annotated centromeres\n",
    "for i in data_nrp1.index:\n",
    "    \n",
    "    if type(data_nrp1.loc[i,'Feature_type'])!=str:\n",
    "        \n",
    "        if data_nrp1.loc[i,'Feature_type'].any() =='Centromere': \n",
    "            \n",
    "            ax2.vlines(x=i,ymin=0,ymax=0.8,linestyles='--',alpha=0.3)\n",
    "            ax2.text(x=i,y=0.6,s='centromere',rotation=90,fontsize=8)\n",
    "    else:\n",
    "\n",
    "         if data_nrp1.loc[i,'Feature_type'] =='Centromere': \n",
    "            \n",
    "            ax2.vlines(x=i,ymin=0,ymax=0.8,linestyles='--',alpha=0.3)\n",
    "            ax2.text(x=i,y=0.6,s='centromere',rotation=90,fontsize=8)\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "##  Plot reads per transposon  highlighting the centromere position"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "strain=data_nrp1\n",
    "\n",
    "fig=plt.figure(figsize=(15,15))\n",
    "grid = plt.GridSpec(4, 1, wspace=0.0, hspace=0.1)\n",
    "\n",
    "ax = plt.subplot(grid[0,0])\n",
    "ax2 = plt.subplot(grid[2,0])   \n",
    "ax3 = plt.subplot(grid[3,0]) \n",
    "ax4 = plt.subplot(grid[1,0])  \n",
    "\n",
    "ax.set_title('Variation along the genome for dnrp1 merged dataset')\n",
    "ax.plot(strain['reads-per-tr'],\".\",alpha=0.9,color='green')\n",
    "ax.set_ylabel('reads per tr filtered for high tr-density')\n",
    "\n",
    "ax.set_ylim(0,2000)\n",
    "\n",
    "ax4.plot(strain['Nreadsperinsrt'],\".\",alpha=0.5,color='green')\n",
    "ax4.set_ylabel('reads per tr without density filter')\n",
    "ax4.set_ylim(0,2000)\n",
    "     \n",
    "\n",
    "ax2.plot(strain['tr-density'],\".\",alpha=0.7,color='black')\n",
    "ax2.set_ylabel('transposon density')\n",
    "ax2.set_ylim(0,1)\n",
    "\n",
    "ax3.plot(strain['reads-density'], \".\",alpha=0.5,color='b')\n",
    "ax3.set_ylabel('Reads density')\n",
    "ax3.set_xlabel('genomic regions')\n",
    "ax3.set_ylim(0,250)\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "source": [
    "#%% Compare the fold change of the reads per transposon per library\n",
    "\n",
    "fold_change=data_wt['reads-per-tr']/data_nrp1['reads-per-tr']\n",
    "fold_change.replace([np.inf, -np.inf], np.nan, inplace=True)\n",
    "fold_change.fillna(0,inplace=True)\n",
    "\n",
    "fold_change_nrp1=data_nrp1['reads-per-tr']/data_wt['reads-per-tr']\n",
    "fold_change_nrp1.replace([np.inf, -np.inf], np.nan, inplace=True)\n",
    "fold_change_nrp1.fillna(0,inplace=True)\n",
    "\n",
    "cutoff=8\n",
    "\n",
    "fig=plt.figure(figsize=(10,30))\n",
    "grid = plt.GridSpec(3, 1, wspace=0.0, hspace=0.2)\n",
    "\n",
    "ax = plt.subplot(grid[0,0])\n",
    "ax1 = plt.subplot(grid[1,0])\n",
    "ax2 = plt.subplot(grid[2,0])\n",
    "\n",
    "ax.set_title('Potential Negative Interactors for nrp1')\n",
    "ax.plot(fold_change,alpha=0.6,color='red')\n",
    "ax.hlines(y=cutoff,xmin=0,xmax=14000,linestyles='--',alpha=0.3,label='cutoff')\n",
    "\n",
    "ax.set_ylabel('fold change reads per tr')\n",
    "ax.set_ylim(0,100)\n",
    "## annotated centromeres\n",
    "for i in np.arange(0,len(data_wt)):\n",
    "    \n",
    "    \n",
    "    if fold_change[i]>cutoff and data_wt.loc[i,'Standard_name']!='noncoding':\n",
    "        ax.vlines(x=i,ymin=0,ymax=100,linestyles='-',alpha=0.2)\n",
    "        ax.text(x=i,y=60,s=data_wt.loc[i,'Standard_name'],rotation=90,fontsize=8)\n",
    "    if fold_change[i]>cutoff and data_wt.loc[i,'Essentiality']==1 :\n",
    "        ax.vlines(x=i,ymin=0,ymax=100,linestyles='-',alpha=0.2)\n",
    "        ax.text(x=i,y=60,s=data_wt.loc[i,'Standard_name'],rotation=90,fontsize=8,color='red')\n",
    "\n",
    "ax1.set_title('Potential Positive Interactors for nrp1')\n",
    "ax1.plot(fold_change_nrp1,alpha=0.6,color='green')\n",
    "ax1.hlines(y=cutoff,xmin=0,xmax=14000,linestyles='--',alpha=0.3,label='cutoff')\n",
    "ax1.set_xlabel('genomic regions')\n",
    "ax1.set_ylabel('fold change reads per tr')\n",
    "ax1.set_ylim(0,100)\n",
    "## annotated centromeres\n",
    "for i in np.arange(0,len(data_wt)):\n",
    "    \n",
    "    \n",
    "    if fold_change_nrp1[i]>cutoff and data_wt.loc[i,'Standard_name']!='noncoding' :\n",
    "        ax1.vlines(x=i,ymin=0,ymax=100,linestyles='-',alpha=0.2)\n",
    "        ax1.text(x=i,y=60,s=data_wt.loc[i,'Standard_name'],rotation=90,fontsize=8)\n",
    "    if fold_change_nrp1[i]>cutoff and data_wt.loc[i,'Essentiality']==1 :\n",
    "        ax1.vlines(x=i,ymin=0,ymax=100,linestyles='-',alpha=0.2)\n",
    "        ax1.text(x=i,y=60,s=data_wt.loc[i,'Standard_name'],rotation=90,fontsize=8,color='red')\n",
    "        \n",
    "ax.legend()\n",
    "ax1.legend()\n",
    "\n",
    "ax2.plot(data_wt['reads-per-tr'],data_nrp1['reads-per-tr'],'o',color='gray')\n",
    "ax2.set_xlim(0,3000)\n",
    "ax2.set_ylim(0,3000)\n",
    "ax2.set_xlabel('reads per tr WT')\n",
    "ax2.set_ylabel('reads per tr dnrp1')"
   ],
   "outputs": [],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.7.10",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.7.10 64-bit ('base': conda)"
  },
  "interpreter": {
   "hash": "b2b6faf28e54db84b8bab82dacb459349e5fe486b652702d3c0dc9e6aff947a2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}